File: LeetCode/Explore/TopInterview/Blind75/LinkedList/19_remove_nth_node_from_end_of_list.cpp
// One pass algorithm

// Complexity Analysis
// Time complexity : O(L)O(L).
// The algorithm makes one traversal of the list of LL nodes. Therefore time complexity is O(L)O(L).
// Space complexity : O(1)O(1).
// We only used constant extra space.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode *f = head, *s = head;
    while (n) {
      f = f->next;
      n--;
    }
    if (!f) {
      return head->next;
    }
    else {
      while (f->next) {
        f = f->next;
        s = s->next;
      }
      s->next = s->next->next;
      return head;
    }
  }
};

File: LeetCode/Explore/TopInterview/Blind75/LinkedList/19_remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        f, s = head, head
        
        while (n):
          f = f.next
          n -= 1
          
        if f == None:
          return head.next
        else:
          while f.next:
            f = f.next
            s = s.next
          
        s.next = s.next.next
        return head  
